File: Sources/WasmVMCore/NinePCodec.swift
import Foundation

/// 9P2000.L wire format per `spec/04-ninep-server.md`.
///
/// Message: `size(4 LE) | type(1) | tag(2 LE) | body`
/// `size` includes header. One WebSocket binary message = one 9P message.

public enum NinePOp: UInt8, Sendable {
    case Tlerror   = 6,  Rlerror   = 7
    case Tstatfs   = 8,  Rstatfs   = 9
    case Tlopen    = 12, Rlopen    = 13
    case Tlcreate  = 14, Rlcreate  = 15
    case Tsymlink  = 16, Rsymlink  = 17  // ENOSYS in MVP
    case Tmknod    = 18, Rmknod    = 19  // ENOSYS in MVP
    case Trename   = 20, Rrename   = 21  // deferred; MVP uses Trenameat (74/75) if needed
    case Treadlink = 22, Rreadlink = 23  // ENOSYS in MVP
    case Tgetattr  = 24, Rgetattr  = 25
    case Tsetattr  = 26, Rsetattr  = 27
    case Treaddir  = 40, Rreaddir  = 41
    case Tfsync    = 50, Rfsync    = 51
    case Tlock     = 52, Rlock     = 53  // ENOSYS in MVP
    case Tmkdir    = 70, Rmkdir    = 71
    case Tunlinkat = 72, Runlinkat = 73
    case Trenameat = 74, Rrenameat = 75  // post-MVP
    case Tversion  = 100, Rversion = 101
    case Tattach   = 104, Rattach  = 105
    case Twalk     = 110, Rwalk    = 111
    case Tread     = 116, Rread    = 117
    case Twrite    = 118, Rwrite   = 119
    case Tclunk    = 120, Rclunk   = 121
}

/// A decoded 9P message header + opaque body.
/// The interpretation of `body` is op-specific; codec helpers parse per-op.
public struct NinePMessage: Sendable, Equatable {
    public let op: NinePOp
    public let tag: UInt16
    public let body: Data

    public init(op: NinePOp, tag: UInt16, body: Data) {
        self.op = op
        self.tag = tag
        self.body = body
    }
}

public enum NinePCodecError: Error, Equatable, Sendable {
    case shortHeader            // < 7 bytes
    case truncatedMessage       // size field claims more bytes than buffer holds
    case unknownOp(UInt8)
    case invalidString          // string field length exceeds remaining bytes or non-UTF-8
    case invalidBody(reason: String)
}

/// 9P qid: `type(1) | version(4 LE) | path(8 LE)`
public struct Qid: Sendable, Equatable {
    public enum Kind: UInt8, Sendable {
        case file    = 0x00
        case symlink = 0x02
        case dir     = 0x80
    }
    public let kind: Kind
    public let version: UInt32
    public let path: UInt64

    public init(kind: Kind, version: UInt32, path: UInt64) {
        self.kind = kind
        self.version = version
        self.path = path
    }
}

public enum NinePCodec {
    public static let headerSize = 7
    public static let qidSize = 13

    public static func encode(_ message: NinePMessage) -> Data {
        let total = headerSize + message.body.count
        var out = Data()
        out.reserveCapacity(total)
        out.appendU32LE(UInt32(total))
        out.appendU8(message.op.rawValue)
        out.appendU16LE(message.tag)
        out.append(message.body)
        return out
    }

    public static func decode(_ bytes: Data) throws -> NinePMessage {
        guard bytes.count >= headerSize else { throw NinePCodecError.shortHeader }
        let size = bytes.readU32LE(at: 0)!
        guard Int(size) == bytes.count else {
            throw NinePCodecError.truncatedMessage
        }
        let opRaw = bytes.readU8(at: 4)!
        guard let op = NinePOp(rawValue: opRaw) else {
            throw NinePCodecError.unknownOp(opRaw)
        }
        let tag = bytes.readU16LE(at: 5)!
        let bodyLen = Int(size) - headerSize
        let body = bytes.slice(at: headerSize, length: bodyLen) ?? Data()
        return NinePMessage(op: op, tag: tag, body: body)
    }

    public static func appendString(_ s: String, to out: inout Data) {
        let utf8 = Data(s.utf8)
        out.appendU16LE(UInt16(utf8.count))
        out.append(utf8)
    }

    public static func readString(in bytes: Data, at offset: Int) throws -> (String, Int) {
        guard let len = bytes.readU16LE(at: offset) else {
            throw NinePCodecError.invalidString
        }
        let strStart = offset + 2
        let strEnd = strStart + Int(len)
        guard strEnd <= bytes.count else {
            throw NinePCodecError.invalidString
        }
        let strData = bytes.slice(at: strStart, length: Int(len)) ?? Data()
        guard let s = String(data: strData, encoding: .utf8) else {
            throw NinePCodecError.invalidString
        }
        return (s, strEnd)
    }

    public static func appendQid(_ qid: Qid, to out: inout Data) {
        out.appendU8(qid.kind.rawValue)
        out.appendU32LE(qid.version)
        out.appendU64LE(qid.path)
    }

    public static func readQid(in bytes: Data, at offset: Int) throws -> (Qid, Int) {
        guard offset + qidSize <= bytes.count else {
            throw NinePCodecError.invalidBody(reason: "qid out of bounds")
        }
        let kindRaw = bytes.readU8(at: offset)!
        guard let kind = Qid.Kind(rawValue: kindRaw) else {
            throw NinePCodecError.invalidBody(reason: "qid kind \(kindRaw)")
        }
        let version = bytes.readU32LE(at: offset + 1)!
        let path = bytes.readU64LE(at: offset + 5)!
        return (Qid(kind: kind, version: version, path: path), offset + qidSize)
    }
}

File: webvm-harness/ninep-codec.js
// 9P2000.L wire format per spec/04-ninep-server.md, JS side.
// Used by webvm-9p-mount.js (the JS-side 9P client) only if the W4 investigation
// concludes that we must implement 9P as a `dir`-mount shim (rather than using
// CheerpX's internal 9P support). If CheerpX accepts a real `9p` mount config
// directly, this file may go unused — keep it small and tested anyway.
//
// Message: size(4 LE) | type(1) | tag(2 LE) | body
//
// `qid.path` is u64 on the wire; this codec uses BigInt for that field on the
// JS side because Number cannot represent the full u64 range exactly.

import {
  writeU8, writeU16LE, writeU32LE, writeU64LE,
  readU8, readU16LE, readU32LE, readU64LE,
  checkBounds, viewOf, concat, assertU8, assertU16, assertU32,
} from "./_le.js";

export const OP = Object.freeze({
  Tlerror:    6,  Rlerror:    7,
  Tstatfs:    8,  Rstatfs:    9,
  Tlopen:    12,  Rlopen:    13,
  Tlcreate:  14,  Rlcreate:  15,
  Tgetattr:  24,  Rgetattr:  25,
  Tsetattr:  26,  Rsetattr:  27,
  Treaddir:  40,  Rreaddir:  41,
  Tfsync:    50,  Rfsync:    51,
  Tmkdir:    70,  Rmkdir:    71,
  Tunlinkat: 72,  Runlinkat: 73,
  Tversion: 100,  Rversion: 101,
  Tattach:  104,  Rattach:  105,
  Twalk:    110,  Rwalk:    111,
  Tread:    116,  Rread:    117,
  Twrite:   118,  Rwrite:   119,
  Tclunk:   120,  Rclunk:   121,
});

export const QID_TYPE = Object.freeze({ FILE: 0x00, SYMLINK: 0x02, DIR: 0x80 });

const MSG_HEADER = 7;
const QID_SIZE = 13;

const utf8enc = new TextEncoder();
const utf8dec = new TextDecoder("utf-8", { fatal: true });

/**
 * Encode a message (size header included).
 * @param {number} op
 * @param {number} tag — u16
 * @param {Uint8Array} body
 * @returns {Uint8Array}
 */
export function encodeMessage(op, tag, body) {
  assertU8(op, "encodeMessage.op");
  assertU16(tag, "encodeMessage.tag");
  if (!(body instanceof Uint8Array)) {
    throw new Error("encodeMessage.body: expected Uint8Array");
  }
  const total = MSG_HEADER + body.byteLength;
  if (total > 0xffffffff) {
    throw new Error(`encodeMessage: total size ${total} exceeds u32`);
  }
  const out = new Uint8Array(total);
  const v = viewOf(out);
  writeU32LE(v, 0, total);
  writeU8(v, 4, op);
  writeU16LE(v, 5, tag);
  out.set(body, MSG_HEADER);
  return out;
}

/**
 * Decode a message. Throws on short or truncated input.
 * @param {Uint8Array} bytes
 * @returns {{ op:number, tag:number, body:Uint8Array }}
 */
export function decodeMessage(bytes) {
  if (!(bytes instanceof Uint8Array)) {
    throw new Error("decodeMessage: expected Uint8Array");
  }
  if (bytes.byteLength < MSG_HEADER) {
    throw new Error(`decodeMessage: short header (${bytes.byteLength} < ${MSG_HEADER})`);
  }
  const v = viewOf(bytes);
  const size = readU32LE(v, 0);
  if (size !== bytes.byteLength) {
    throw new Error(`decodeMessage: size mismatch (header says ${size}, frame is ${bytes.byteLength})`);
  }
  const op = readU8(v, 4);
  const tag = readU16LE(v, 5);
  const body = bytes.slice(MSG_HEADER, size);
  return { op, tag, body };
}

/**
 * Encode a 9P string (len(2 LE) | utf8 bytes) appended to `chunks`.
 * @param {string} s
 * @param {Uint8Array[]} chunks — caller accumulates and concatenates
 */
export function appendString(s, chunks) {
  if (typeof s !== "string") {
    throw new Error("appendString: expected string");
  }
  if (!Array.isArray(chunks)) {
    throw new Error("appendString: chunks must be an array");
  }
  const bytes = utf8enc.encode(s);
  if (bytes.byteLength > 0xffff) {
    throw new Error(`appendString: too long (${bytes.byteLength} > 65535)`);
  }
  const header = new Uint8Array(2);
  writeU16LE(viewOf(header), 0, bytes.byteLength);
  chunks.push(header);
  chunks.push(bytes);
}

/**
 * Read a 9P string starting at `offset`. Returns [string, newOffset].
 * @param {Uint8Array} bytes
 * @param {number} offset
 * @returns {[string, number]}
 */
export function readString(bytes, offset) {
  if (!(bytes instanceof Uint8Array)) {
    throw new Error("readString: expected Uint8Array");
  }
  checkBounds(bytes, offset, 2, "readString.length");
  const v = viewOf(bytes);
  const len = readU16LE(v, offset);
  const start = offset + 2;
  checkBounds(bytes, start, len, "readString.body");
  let s;
  try {
    s = utf8dec.decode(bytes.subarray(start, start + len));
  } catch (e) {
    throw new Error(`readString: invalid UTF-8: ${e.message}`);
  }
  return [s, start + len];
}

/**
 * Encode a 13-byte qid: type(1) | version(4 LE) | path(8 LE).
 * @param {{ type:number, version:number, path:bigint }} qid
 * @param {Uint8Array[]} chunks
 */
export function appendQid(qid, chunks) {
  if (qid == null || typeof qid !== "object") {
    throw new Error("appendQid: expected qid object");
  }
  if (!Array.isArray(chunks)) {
    throw new Error("appendQid: chunks must be an array");
  }
  assertU8(qid.type, "appendQid.type");
  assertU32(qid.version, "appendQid.version");
  if (typeof qid.path !== "bigint") {
    throw new Error("appendQid.path: expected BigInt");
  }
  if (qid.path < 0n || qid.path > 0xffffffffffffffffn) {
    throw new Error(`appendQid.path: out of u64 range (${qid.path})`);
  }
  const buf = new Uint8Array(QID_SIZE);
  const v = viewOf(buf);
  writeU8(v, 0, qid.type);
  writeU32LE(v, 1, qid.version);
  writeU64LE(v, 5, qid.path);
  chunks.push(buf);
}

/**
 * Read a qid starting at `offset`. Returns [qid, newOffset].
 * @param {Uint8Array} bytes
 * @param {number} offset
 * @returns {[{type:number, version:number, path:bigint}, number]}
 */
export function readQid(bytes, offset) {
  if (!(bytes instanceof Uint8Array)) {
    throw new Error("readQid: expected Uint8Array");
  }
  checkBounds(bytes, offset, QID_SIZE, "readQid");
  const v = viewOf(bytes);
  const type = readU8(v, offset);
  const version = readU32LE(v, offset + 1);
  const path = readU64LE(v, offset + 5);
  return [{ type, version, path }, offset + QID_SIZE];
}

export { concat };
